internal/scheduler/jobdb.go

```diff
@@ -56,31 +199,91 @@ func NewJobDb() (*JobDb, error) {
 	}, nil
 }
 
-func (jobDb *JobDb) Upsert(jobs []*SchedulerJob) error {
-	txn := jobDb.Db.Txn(true)
-	defer txn.Abort()
+// Upsert will insert the given jobs if they don't already exist or update the if they do
+// Any jobs passed to this function *must not* be subsequently modified
+func (jobDb *JobDb) Upsert(txn *memdb.Txn, jobs []*SchedulerJob) error {
 	for _, job := range jobs {
-		err := txn.Insert("jobs", job)
+		err := txn.Insert(jobsTable, job)
 		if err != nil {
 			return errors.WithStack(err)
 		}
 	}
-	txn.Commit()
 	return nil
 }
 
+// GetById returns the job with the given Id or nil if no such job exists
+// The Job returned by this function *must not* be subsequently modified
```

Comment:
> When you say "must not be subsequently modified" you mean the structs returned by this method must not be modified by callers of this method?

Yes this is correct. It's a requirement of HashiCorp's memDb lib which we are using under the covers.

I like the idea of making `SchedulerJob` immutable (I worked for a long time as a scala dev and I miss things being immyutable by default!) but I don't think it's trivial because `SchedulerJob` has is a relatively complex objec t so it'll be quite a bit of code.  Any  objections if we do this in a follow-up PR once we've got this working E2E?